### ISSUE TITLE
Incorrect variance inference when generic callable attributes/properties present

### ISSUE THREAD

**Issue by randolf-scholz** [microsoft/pyright#7907]
`pyright` misses variance violation when property is annotated as `Callable` or callback-protocol. [[pyright-playground]](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDCAhgDYlEBGJApgDRQDi1K1ISAxvQArgxjtgS9ACqJqANSIh6AZWolgAKEXCoAXiiiEEqQAoARMP0BKFQH0B6zWMkgDwi2H30BANylIiKGGuEgArtSmDgLeIERWWjp2ho5hRM5QoTDh7mxePn6Bpors5ADO%2BVAAcv5kALJECLo8YHwCJADaIWDxALrGAFyKUL1QACbUwFBmFqQko7r58sD0AB6dmnEpRPQA9MZQALQAfCWt1IsAdCfKeUSFUABiYGC6TCxs7M2OHd19A0OYKEgwUzPzRYtdabXb7FjHE69ADEUAgiDgi0AMuT0BBwNhoAAWMGRZwKRQAQnoHqwOC8BG8en0AAIIcDaPBwKm9QbDVC-TwkJAAL1Y-wUoL2xDIlBojXJYDa9GKBzaiygsPhaORqPRSCxOKggFByZkfKAgagwfwgFBQcgQCj9CILcHUPEXQlEbn3Zik54tSkfWn01jwXWs74c0g8vnTAXbPalCpVCVymFwhGLLWqjHYpO6j4Go0ms1EC1WqA2mUse2XACK-jmLseZI9XV17JgnJDIEWwvIVGo4rkChEr2lsqs5stETDs0Li2L1HjSsRUBRUDRqc1WqAA), [[mypy-playground]](https://mypy-play.net/?mypy=latest&python=3.12&flags=strict&gist=c730da8e9cbb9f925a4cbf2f892ca40f)

```python
from typing import Callable, Generic, Protocol, TypeVar, Self

T = TypeVar("T")
T_co = TypeVar("T_co", covariant=True)
T_contra = TypeVar("T_contra", contravariant=True)

class NullMap(Protocol[T_contra]):
    def __call__(self, x: T_contra, /) -> None: ...

class Foo(Generic[T_co]):
    def init(self, x: T_co, /) -> None: ...  # mypy: ❌, pyright: ❌

class Bar(Generic[T_co]):
    @property
    def initializer(self) -> Callable[[T_co], None]:  # mypy: ❌, pyright: ✅
        return lambda x: None

class Baz(Generic[T_co]):
    @property
    def initializer(self) -> NullMap[T_co]:  # mypy: ✅, pyright: ✅
        return lambda x: None

class Qux(Generic[T_co]):
    initializer: Callable[[Self, T_co], None] = lambda self, x: None  # mypy: ❌, pyright: ✅
```

In all 4 cases, an invariant type-variable must be used, because `Callable[[S], None]` is generally not a subtype of `Callable[[T], None]` when `S<:T`. This can lead to LSP violation like so: [[pyright-playground]](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgGoCGIANFAOLErEhIDGNAwuQDZfkBGXYgCghLXgGdxUAIIAuIVEVQAJsWBQQxccRgAKbV2ABKKAFoAfFAByYRrKgA6JyIIB9FmCgBeQiQohdACI3D0CjIWl3Tx8iBDJKIMjQmj4wAFcUZS9pcNEJKQBRAA9ybEFdekZmFgBtELAAXSN5JSgAAQRwOLw4BSVVdVQkGCRuJAAvJn1iQxMLKE4efkEauqiGmhtGBpbWpU0YNJAUKF4IPmVyKFd7LeE88kkoAGU0vmLShHKPsuIapMazT6ig6XSY8GBKjUmBQw1GXAmUwMxjMlkWvAEf3%2B602tmIO0hrQORxOZwuVyK9iKDk02j0uQGUGAYDAumIJV%2B9h%2BXz%2BYD4ACtiCwYE1UdY8btFEVvFA%2BYLhbpwq12Z9BA4hiMxpMAkVcszWa93hyeYqTFAAMS%2BOIFEDgEBQcj4EAZEYQYRAA), [[mypy-playground]](https://mypy-play.net/?mypy=latest&python=3.12&flags=strict&gist=c89ad1b9be67b64876a6eb8fa4db087b)

```python
from typing import TypeVar, Generic, Callable

class A:
    def reset(self) -> None: ...

T_co = TypeVar("T_co")
A_co = TypeVar("A_co", bound=A)

class Example(Generic[T_co]):
    @property
    def initializer(self) -> Callable[[T_co], None]:
        return lambda _: None

class SubExample(Example[A_co]):
    @property
    def initializer(self) -> Callable[[A_co], None]:
        return lambda x: x.reset()

def foo(example: Example[object]) -> None:
    x = object()
    example.initializer(x)

foo(SubExample())  # TypeError at runtime
```

Or, with 3.12 / PEP695 syntax: [[pyright-playground]](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDCAhgDYlEBGJApgFC0DG5Azs1AIIBctUvUAJtWBQQ1ZtRgAKcSWABKKAFoAfFAByYFNU5QAdPvqMWbAKIAPIthoBtACoBdSXO58oAAQTgE1PHB59BYVQkGCRSJAAvH2lqWQUVQlJyKmprO3sAGnVNansXVz5RGABXEBQocggKfiIoAH0dDS16JiJWKABlYopzSwQbDp12R16rVI77Z39eDy8feGmBIUwUELCSSOiZeSVVYjJKG2sJrKbc-IKRCVLyyurasx0zXVFxKTl6QKhgMDBJagsYx0o36qTAFAAVtQGDBJrtsloLlAzFAALxQcFQmFORYAvo0XTBULhKIgSRmD60H5-Lo9QGgpwKKAAYigtkQ1BMIHAICgRHwIGKKFCEDoQA)

```python
from typing import Callable

class A:
    def reset(self) -> None: ...


class Example[T]():
    @property
    def initializer(self) -> Callable[[T], None]:
        return lambda _: None

class SubExample[S: A](Example[S]):
    @property
    def initializer(self) -> Callable[[S], None]:
        return lambda x: x.reset()

def foo(example: Example[object]) -> None:
    x = object()
    example.initializer(x)

foo(SubExample())  # TypeError at runtime
```

**Comment by erictraut** [comment]
I think your analysis is wrong here. The reason you're not seeing a type error in your last sample is because you're instantiating `SubExample` in such a way that its type argument is unsolved, and therefore defaults to `Any`. If you change the last line of your example to `foo(SubExample[A]())`, you'll see that pyright (and mypy) both generate a type error here. Likewise, if you use `foo(SubExample[object]())`, both pyright and mypy generate a type error.

I think pyright's invariance enforcement is correct here.

Mypy's behavior here appears to be buggy. If you replace the bodies of the methods with `...` and do not assign a value to `Qux.initializer`, mypy no longer complains about TypeVar variance.

```python
class Bar(Generic[T_co]):
    @property
    def initializer(self) -> Callable[[T_co], None]: ...

class Baz(Generic[T_co]):
    @property
    def initializer(self) -> NullMap[T_co]: ...

class Qux(Generic[T_co]):
    initializer: Callable[[Self, T_co], None]
```

Adding back the method bodies and the assigned value for `Qux.initializer` should not affect variance enforcement or inference, so there's something wrong going on with mypy. I'll let you decide whether or not you want to report it.

In any case, I think pyright's behavior here is correct. I don't see a bug here, so I'm closing the issue.

**Comment by randolf-scholz** [comment]
Eric, thank you for your feedback. I noticed the different behavior when explicitly annotating the example as well, but I didn't include it because it didn't seem relevant. There was also a typo in the second example (missing `covariant=True` in `TypeVar` definition), so I edited my OP.

If `E[+T]` is covariantly generic, then `E[S] <: E[T]` whenever `S <: T`. In particular, if `F <: E[S]` for *some* `S <: T`, then by transitivity `F <: E[T]`. Applied to the example above:

1. `Example[A] <: Example[object]` (covariance of `Example[T_co]`)
2. `SubExample[A] <: Example[A]`
3. ∴ `SubExample[A] <: Exmaple[object]`.

Here's an updated example that shows the problem, and without forgetting `covariant=True`, `mypy` finds this issue but pyright doesn't [[mypy-playground]])(https://mypy-play.net/?mypy=latest&python=3.12&gist=5c300bd14ec49f6152a1fc31e7a6edd4),  [[pyright playground]](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgGoCGIANFAOLErEhIDGNAwuQDZfkBGXYgCghLXgGdxUAIIAuIVEVQAJsWBQQxccRgAKbV2ABKKAFoAfFAByYRrKgA6JyIIB9FmCgBeQiQohdACI3D0CaDwA3SiRyFBgvAhAAV2IjIWl3Tx8iBDJKIIzQmj4wJJRlL2lwsCjmWPjElLTRCSkAUQAPcmxBXXpGZhYAbRCwAF0jeSUoAAEEcFy8OAUlVXVUJBgYriQALyZ9YkMTCyhOHn5BIZHMsZobRjH7KABiKAhEOHtAGXIaBDhmGgABYweyAUHIVtNFJoYEkQCgoLwIHxlOQoK57A9hC1yJIoABlJJ8TrdBC9Ek9YhDQrjSaQ2bzMCLeD0taYFCbbZ7A4GYxmSznXgCKnU273WzEJ6KN4ff4-P4ApDA0FQCFQpQwuEIpEotEdewdByabR6ZpspIIFi4mCuYhdSm6O2kwT2ClkqnSCb8qBuq5gPgAK2ILBgT3pmvhUCdlJE5st1tc4iJ0fd%2BmT9vd9kJxIzV09J0svo9YemEYRSb4KcEsbUUGAYDAjtzxFdzaG-qDIa9pyxUyUHW8UA7wdN9KrxAcGy23G5AQ6zSEFfHWaJRepY0H2aLujS49cch9bc9g4tVvENqXzbTlebu%2BbrmHIdbzqpj9DJ-j59tV730jS9cbPc3xMV5fFyNoQHAEAoHIfBkjiLBhCAA)

```python
from typing import TypeVar, Generic, Callable

class A:
    def reset(self) -> None: ...

T_co = TypeVar("T_co", covariant=True)
A_co = TypeVar("A_co", bound=A, covariant=True)

class Example(Generic[T_co]):
    @property
    def initializer(self) -> Callable[[T_co], None]:  # mypy: ❌, pyright: ✅
        return lambda _: None

class SubExample(Example[A_co]):
    @property
    def initializer(self) -> Callable[[A_co], None]:  # mypy: ❌, pyright: ✅
        return lambda x: x.reset()

def upcast_example(example: Example[A]) -> Example[object]:
    return example

def upcast_subexample(subexample: SubExample[A]) -> Example[A]:
    return subexample

def foo(example: Example[object]) -> None:
    x = object()
    example.initializer(x)


subexample: SubExample[A] = SubExample()
example_A: Example[A] = upcast_subexample(subexample)
example_object: Example[object] = upcast_example(example_A)
foo(example_object)  # TypeError at runtime
```

**Comment by erictraut** [comment]
Neither pyright nor mypy detect class-scoped type variables that are defined with incorrect variance in the general case. This is done only for protocol classes, not for regular classes. If you change the classes in your example to protocol classes, pyright will generate a diagnostic telling you that the type variable should be contravariant. So will mypy.

Both mypy and pyright do have some basic checks in place to prevent obvious misuse of covariant and contravariant type variables in method parameter annotations and return types within non-protocol classes, but these checks are not comprehensive. This is one of the reasons I pushed for automatic variance calculation in PEP 695. It's difficult for humans to get variance correct, especially complex cases.

In mypy's case, these basic checks appear to be buggy. The fact that it's generating an error in your example is probably not what the mypy authors intended. As evidence of this, you can see that the error goes away if you replace the `return` statements with `...`. (The returned value clearly shouldn't affect this diagnostic if it were working correctly.) The diagnostic also goes away if you change the return type from a `Callable` to a union between a `Callable` and some other type, such as `None`.

In short, I don't consider this a bug in pyright. It's working as intended. If you want to avoid errors like this, I recommend switching to PEP 695 syntax.
